### ISSUE TITLE
Importing request-to-curl leaks memory on each request

### ISSUE THREAD

**Issue by nicolasazrak** [JarvusInnovations/node-request-to-curl#1]
After using this library I have noticed that my application started using more memory every time. I could isolate the problem and it came from this module. It can be reproduced with the following code: 

```js
require('request-to-curl');
const request = require('request');
const Bluebird = require('bluebird');

const initial = process.memoryUsage();
Bluebird
  .mapSeries(new Array(50000), function() {
    return new Bluebird((resolve) => {
      return request('http://localhost:8080', function(err, response) {
        response.request.req.toCurl();
        resolve();
      });
    });
  })
  .then(() => {
    global.gc();
    const final = process.memoryUsage();
    console.log(`Diff RSS = ${(final.rss - initial.rss) / 1024 / 1024} MB`);
    console.log(`Diff HEAP = ${(final.heapUsed - initial.heapUsed) / 1024 / 1024} MB`);
  })
  .catch(e => {
    console.error(e);
  });
```

It mus be run with `node --expose-gc`. After 50.000 iterations, RSS has grown `356.71875 MB`. Commenting the `require` and `toCurl()` solves the problem.

**Comment by damianmr** [comment]
happens to me as well 👎

**Comment by aqiongbei** [comment]
I have the same question!

**Comment by rsilvestre** [comment]
Same here. If you read this thread and you want to use this module in production.

**MEMORY LEAKS. DON'T USE THIS MODULE**

**Comment by themightychris** [comment]
The biggest potential culprit for accumulating memory I can see is the storage of request bodies in the request object instances. That's necessary for being able to `.toCurl()` later though.

So is this just about the request instances being garbage collected? References to them get stashed in the response object. The test code above discards the responses array in its handler, but that array of responses is probably still in scope in the promise library during the handler being run

I'm mobile right now but curious what the results of the above test would be if setTimeout was used inside the request's `then` handler with some delay to allow the promise call stack to close up and the promise get GCd.

If it is just a matter of accumulating request instances in user code, then this library is just multiplying the effect of an existing memory leak, proportional to request body sizes.

Are others seeing this issue using substantial request bodies? Anyone noticing it while only doing GET requests?

**Comment by lincanbin** [comment]
MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added. Use emitter.setMaxListeners() to increase limit

**Comment by themightychris** [comment]
We're not currently maintaining any instances of the application we used this in, can anyone take a stab at a pull request?

**Comment by jmealo** [comment]
Hello everyone! Original author here, I don't get notifications on this repo; sorry for the slow response.

The `http-parser` in Node.js core leaks memory under certain conditions, it's being replaced by `llhttp`. If you switch out `http-parser` (native/in-core) with `http-parser-js` that *mostly* fixes the problem.

When a socket is used for multiple requests by the http agent (`keepalive: true`), this module attached a `close` event listener and wrapped `socket.write` and `socket.ondata` on each request (rather than once per-socket; which was the assumption this code made). Under sufficient load a stack overflow was possible (though not reported).

@nicolasazrak Thanks for the snippet, this made it easy for me to verify that something was wrong. I assumed the error was in my code and that`http-parser` was safe to use. It wasn't until I reproduced the issues reported by @lincanbin and swapped out `http-parser` for `http-parser-js` that I realized there were two separate unrelated memory leaks and that my code didn't contain logic errors, it works fine with `keepalive: false`, it just makes a bad assumption,  which every time I tried to make an alternate implementation to fix the issue (using weakmaps, buffers versus strings, closure versus bind) the leaks still occurred. I assume that my usage of `http-parser` is correct as it does not leak when used with `http-parser-js`.

I'm sorry for the pain that this has caused anyone! I refrain from using this or any npm modules that do not have unit tests and always check the issues (@nicolasazrak  @damianmr  @aqiongbei  @rsilvestre @lincanbin  you saved countless others, thank you for taking the time to report this issue!). In all honesty, my tests probably wouldn't have caused these conditions and hopefully we're all testing with realistic load prior to production deployments.

To prevent further chaos/pain: **I will be deprecating the versions published to NPM with warnings about a memory leak.** I'm currently assembling some test cases to verify that this module behaves well up to a certain request body size so that I can provide guidance  in the README on when this module should not be used. The bold assertion that this was safe to use in production did not take into account load and is regrettable. 

If I had to redo this, I'd consider doing it at the HTTP Agent level which would make it easy to isolate the memory amplification required to make this work. When the http agent changed it's behavior from re-using sockets based on hostname:port to a more general approach I believe this module (as published) may behave worse with later versions of Node.js.

@themightychris I would like to retire this repo and point at mine since I'm maintaining it. I think/hope that I published this with an npm account that I have access to :).
